### PR TITLE
Drop nil from Vec, BitSet, String, Exception and Error

### PR DESCRIPTION
**Pull request by soc**
Additional pieces extracted from https://github.com/dinfuehr/dora/pull/128.

### CODE REVIEW

**Comment by soc** [comment]
After this PR there are no nils in the standard library left with the exception of https://github.com/dinfuehr/dora/pull/128/files#diff-fdd12e40f618e7564137f156b2f8ba69, which fails due to some bug in initialization.

File: dora/src/object.rs [hunk 1]
```diff
@@ -517,7 +521,7 @@ impl Str {
 
             handle
         } else {
-            Ref::null()
```
Comment (dinfuehr) [inline dora/src/object.rs]: I don't think we should silently swallow this, better to throw an exception here or make the return type `Option[String]`.

File: dora/src/object.rs [hunk 2]
```diff
@@ -493,7 +497,7 @@ impl Str {
         let total_len = val.len();
 
         if offset > total_len {
-            return Ref::null();
```
Comment (dinfuehr) [inline dora/src/object.rs]: Here as well

**Comment by dinfuehr** [comment]
I think we should keep nil in those two places in this PR and fix it in a subsequent one.

**Comment by dinfuehr** [comment]
Wondering about the initialization bug: Do you have a repro?

File: dora/src/object.rs (hunk 1 above)
Comment (soc) [reply dora/src/object.rs]: This is handled in https://github.com/dinfuehr/dora/pull/152/files#diff-edf3ee5de6272371b58c0a2d02e7c112.

File: dora/src/object.rs (hunk 2 above)
Comment (soc) [reply dora/src/object.rs]: This too: https://github.com/dinfuehr/dora/pull/152/files#diff-edf3ee5de6272371b58c0a2d02e7c112.

File: dora/src/object.rs (hunk 1 above)
Comment (dinfuehr) [reply dora/src/object.rs]: Right sorry, haven't seen that... Couldn't we throw the exception directly from the native code instead? Then we wouldn't need the wrappers (fromStringPart and fromBytesPart) anymore and fromStringPartOrNull and fromBytesPartOrNull can simply replace them.

**Comment by dinfuehr** [comment]
I am also wondering whether we should add special "magic" to the `Option` type in the runtime/compiler, such that `Option[SomeClass]` is already just the reference without the indirection and performance overhead. This also would make it easier to use Option types in native Rust code. We could remove part of this magic later when we get ADTs later.

**Comment by soc** [comment]
> I am also wondering whether we should add special "magic" to the Option type in the runtime/compiler, such that Option[SomeClass] is already just the reference without the indirection and performance overhead. This also would make it easier to use Option types in native Rust code. We could remove part of this magic later when we get ADTs later.

I absolutely agree with this (though I'm not qualified to implement it). :-)

File: dora/src/object.rs (hunk 1 above)
Comment (soc) [reply dora/src/object.rs]: Ah, sorry, I just migrated to native excpetions + options before reading this ... I'll push and then you can decide if it's ok or needs to be reworked.

Comment (soc) [reply dora/src/object.rs]: Does this look ok? https://github.com/dinfuehr/dora/pull/152/commits/77e8b3c5fd69f4f9fbdf87565d48b1dc65b0ddae

Comment (dinfuehr) [reply dora/src/object.rs]: Sure, even better ;)

**Comment by dinfuehr** [comment]
> > I am also wondering whether we should add special "magic" to the Option type in the runtime/compiler, such that Option[SomeClass] is already just the reference without the indirection and performance overhead. This also would make it easier to use Option types in native Rust code. We could remove part of this magic later when we get ADTs later.
> 
> I absolutely agree with this (though I'm not qualified to implement it). :-)

Me too, I have no idea how hard this is going to be. I'll give it a try.

Very cool, thanks!

File: dora/src/object.rs (hunk 1 above)
Comment (soc) [reply dora/src/object.rs]: Ah, great! Thanks!

**Pull request merged**